src/Npgsql/NpgsqlSchema.cs

```diff
@@ -563,6 +566,247 @@ static DataTable GetDataSourceInformation(NpgsqlConnection conn)
+                    row["ProviderDbType"] = (int)baseMapping.NpgsqlDbType.Value;
+            }
+
+            return table;
+        }
+
+        /// <summary>
+        /// Populates some generic type information that is common for base types, arrays, enums, etc. Some will
+        /// be overridden later.
+        /// </summary>
+        static void PopulateDefaultDataTypeInfo(DataRow row, PostgresType type)
+        {
+            row["TypeName"] = type.DisplayName;
+            // Skipping ColumnSize at least for now, not very meaningful
+            row["CreateFormat"] = type.DisplayName.ToUpperInvariant();
+            row["CreateParameters"] = "";
+            row["IsAutoIncrementable"] = false;
+            // We populate the DataType above from mapping.ClrTypes, which means we take the .NET type from
+            // which we *infer* the PostgreSQL type. Since only a single PostgreSQL type gets inferred from a given
+            // .NET type, we never have the same DataType in more than one row - so the mapping is always the
+            // best match. See the hardcoding override  below for some exceptions.
+            row["IsBestMatch"] = true;
+            row["IsCaseSensitive"] = true;
+            row["IsConcurrencyType"] = false;
+            row["IsFixedLength"] = false;
+            row["IsFixedPrecisionAndScale"] = false;
+            row["IsLiteralSupported"] = false;  // See hardcoding override below
+            row["IsLong"] = false;
+            row["IsNullable"] = true;
+            row["IsSearchable"] = true;
+            row["IsSearchableWithLike"] = false;
+            row["IsUnsigned"] = DBNull.Value; // See hardcoding override below
+            // LiteralPrefix: no literal for now except for strings, see hardcoding override below
+            // LiteralSuffix: no literal for now except for strings, see hardcoding override below
+            row["MaximumScale"] = DBNull.Value;
+            row["MinimumScale"] = DBNull.Value;
+            // NativeDataType is unset
+            row["OID"] = type.OID;
+        }
+
+        /// <summary>
+        /// Sets some custom, hardcoded info on a DataType row that cannot be loaded/inferred from PostgreSQL
+        /// </summary>
+        static void PopulateHardcodedDataTypeInfo(DataRow row, PostgresType type)
+        {
+            if (type.Name == "text" || type.Name == "varchar" || type.Name == "char")
+            {
+                row["CreateFormat"] += "({0})";
+                row["CreateParameters"] = "size";
+                row["IsSearchableWithLike"] = true;
+                row["IsLiteralSupported"] = true;
+                row["LiteralPrefix"] = "'";
+                row["LiteralSuffix"] = "'";
+
+                if (type.Name == "varchar" || type.Name == "char")
+                {
+                    row["DataType"] = "String";
+                    row["IsBestMatch"] = false;
+                }
+            }
+
+            if (type.Name == "numeric")
```

Comment:
If I didn't overlook something, all the following `if`s check `type.Name` and are mutually exclusive.
How about putting an `else` in front of all them or even convert all of this into a `switch` statement (personally I find `switch` statements easier to read than a series of `if`statements that all care about the same value).
IMO this would make the code clearer and eliminate a few unnecessary equality checks.

Reply:
They're actually not mutually exclusive (numeric and xid participate in two blocks), but I tried and it still looks better/more readable even with a bit of duplication. Thanks.